core/include/vecmem/memory/resources/memory_resource.hpp

```diff
@@ -14,7 +14,14 @@
  * std::experimental::pmr namespace depending on the GCC version used, so we try
  * to unify them by aliassing depending on the compiler feature flags.
  */
-#if __has_include(<memory_resource>)
+#ifdef __CUDACC__
+namespace vecmem {
+    class memory_resource;
+
+    template<typename T>
+    class polymorphic_allocator;
+}
+#elif __has_include(<memory_resource>)
```

Comment:
:thinking: I wonder if we couldn't make this a bit more robust...

How about introducing a `memory_resource_fwd.hpp` file that only has these forward declarations? And then we use that everywhere where we don't need to see the full definitions. Regardless whether CUDA or whatever is being used for the compilation.

In the `jagged_vector_view.hpp` file then you could also hide the constructor needing this type using `__cplusplus` instead of `__CUDACC__`.

I'm just very much on the mind now, thanks to you, that depending on the language specific macros should be avoided as much as possible...

Reply:
We could alternatively have the check be `#if __cplusplus < 201703L` which might be more robust. Would that be okay for you?

Reply:
I don't think that would work in this file. :thinking:

My problem is that I don't like that a given header would either provide the full definition of a type or just a forward declaration of it, based on some circumstances. I think that's not a good design. :frowning:

I believe it would be cleaner if we coded this up like:

`jagged_vector_view.hpp`:
```c++
#include "vecmem/memory/resources/memory_resource_fwd.hpp"
...
      jagged_vector_view( memory_resource& resource );
...
```

`jagged_vector_view.ipp`
```c++
#if __cplusplus >= 201703L
#   include "vecmem/memory/resources/memory_resource.hpp"
#endif
...
template<...>
jagged_vector_view<...>::jagged_vector_view( memory_resource& resource ) {
#if __cplusplus >= 201703L
   // Do whatever needs to be done.
#else
   static_assert( false, "This constructor is not available with <C++17" );
#endif
}
```

Reply:
I understand your hesitation to conditionally provide a forward declaration, but unfortunately I don't think your solution would work either. Typedeffing a class after forward-declaring it is a compilation error, so if we were to go this way we would need to make absolutely sure that no file ever includes, directly or indirectly, both `memory_resource.hpp` and `memory_resource_fwd.hpp`. That makes developing the library as well as using it really difficult. :disappointed:

And I really do agree with you that I think this is not great, but it's the only way I have come up with so far that makes my jagged vector code work.

Reply:
Okay... How about we switch to declaring `vecmem::memory_resource` like this then?

```c++
namespace vecmem {
   class memory_resource : public std::pmr::memory_resource {
   public:
      using std::pmr::memory_resource::memory_resource;
   };
}
```

(With the additional logic for "older" compilers of course.) Then we could freely forward declare it as we'd like.

Reply:
The problem with that is that we would have to do the same for `std::pmr::polymorphic_allocator` which, unlike `std::pmr::memory_resource`, is not a virtual class designed to be subclassed. And we all know that inheriting from `std` classes not designed for that is a dangerous idea.

Reply:
I guess we could split the `memory_resource.hpp` file in two, one for the `memory_resource` class and one for the `polymorphic_allocator` class. I can give it a try, but I worry we might be getting ourselves in a really hairy situation.

Reply:
I was wondering about that since a while as well. I understand that since STL put both types into [<memory_resource>](https://en.cppreference.com/w/cpp/header/memory_resource), you did the same in our header.

For following the ATLAS coding guidelines, I would actually prefer to have 2 separate headers for `vecmem::memory_resource` and `vecmem::polymorphic_allocator<...>`...

Could you give it a try? I agree that it's not a super-clean, super-nice setup. But I really fear adding conditional compilation into the core library based on `__CUDACC__`. (The types in `types.hpp` don't count. :stuck_out_tongue:)